### ISSUE TITLE
Fully working and clean Leo on PyPi.org

### ISSUE THREAD

**Issue by maphew** [leo-editor/leo-editor#562]
## Background

PyPi is the [Python Package Index Page](https://pypi.python.org/pypi)

There has been a lengthy [discussion](https://groups.google.com/d/topic/leo-editor/yLCtg54aqEI/discussion,https://groups.google.com/d/topic/leo-editor/) of this issue on leo-editor.

Leo developers *should not use pip* because some files left behind when using pip (see #603). Instead, Leo devs should use git as always.

## Goals  

- Any Leo maintainer can update [Leo's PyPi page](https://pypi.python.org/pypi/leo) with a straightforward workflow. Expertise with setuptools or distutils expertise should not be required.
- People can install Leo with `pip install leo`.
- People may also have to install PyQt with their favorite package manager: `conda install pyqt`.

## What Matt has done

- Cherry picked the old setup.py branch into a new branch, setuppy-clean, uploaded the result to PyPi, and tested the resulting PyPi package on Windows and Linux.
- Merged the new into master with a pull request and deleted the setuppy-clean branch.
- Updated the docs and scripts in `leo/dist/leoDist.leo`.

## Current status

- Work is now being carried out in master.
- PyPi compliant wheel packages are built and install successfully on Windows (7,10) and Linux (Mint) with python 2.7 and 3.6.
- leoDist.leo contains the **package upload recipe** in the node `Matt's workflow for updating pypi`.  At present, a clone of this node appears at the top level.  The clone will remain at the top level until this issue is closed, or perhaps even longer.

@maphew  **I have removed the package upload recipe from this comment**.  I see no reason for the duplication.  It will be easy enough to restore if necessary.

## To do

- [ ] Update installation docs. **In Progress**. @maphew Please put the various documentation sections in leoDist.leo into separate nodes, with subsidiary discussions in child nodes, or even grandchild nodes.  That way devs can see the logical relationships between sections.  In general, **never** rely on markup when outline structure is available.

- [ ] Someone besides Matt runs through process from start to finish and is successful. @maphew Is further "internal" testing really necessary?  Can't we simply test the wheel after you upload it?

**Comment by edreamleo** [comment]
Thanks for your help on this issue.

**Comment by maphew** [comment]
As far as I'm concerned everything of value in the setup.py branch has been retrieved and it can now be deleted. Any reason not to do that now @edreamleo, @tbnorth , @vivainio ?

**Comment by maphew** [comment]
Steps 1 and 2 are complete. 

   pip install --pre --index-url https://test.pypi.org/simple/ leo 

Works on Windows 7 py2.7.14 and Linux Mint 18.2 (py2.7.12).

**Comment by maphew** [comment]
I think this branch is feature complete. Before we can upload to pypi properly this branch needs to be merged with master so the new setup.py and manifest.in are available, and then a new release tag created, e.g. 5.6.1. It's used like:

    git checkout tags/5.6.1
    python setup.py bdist_wheel
    twine upload --repository-url https://pypi.org/legacy/ dist/*
    pip install leo

I'm pushing the install docs update step to later.

**Comment by tbnorth** [comment]
On Thu, 26 Oct 2017 22:20:20 -0700
matt wilkie <notifications@github.com> wrote:

> As far as I'm concerned everything of value in the setup.py branch
> has been retrieved and it can now be deleted. Any reason not to do
> that now @edreamleo, @tbnorth , @vivainio ?

No objections from me - this isn't something I've worked on.  Great
work Matt, a pip install for Leo is something we should definitely have.

Cheers -Terry

**Comment by edreamleo** [comment]
On Fri, Oct 27, 2017 at 7:32 AM, Terry Brown <notifications@github.com>
wrote:

> On Thu, 26 Oct 2017 22:20:20 -0700
> matt wilkie <notifications@github.com> wrote:
>
> > As far as I'm concerned everything of value in the setup.py branch
> > has been retrieved and it can now be deleted. Any reason not to do
> > that now @edreamleo, @tbnorth , @vivainio ?
>
> No objections from me - this isn't something I've worked on. Great
> work Matt, a pip install for Leo is something we should definitely have.
>

​Yes, many thanks for this work.

I see that most changes involve setup.py, setup.cfg and leoDist.leo.  These
changes clearly can be merged into master.

I do have one question as the result of looking at the diffs.  At rev
80726ff2 you added this line to manifest.in:

​    prune leo/extensions

This seems dubious to me. The leo/extensions folder is important.

Edward

**Comment by maphew** [comment]
The prune statements were migrated from the setup.py of several years ago, https://github.com/leo-editor/leo-editor/blame/9982d17e2470ac8313050b8a0288cd39d7ad4f5b/setup.py#L62. With zero thinking of why, I put them in because they were excluded before. Looking at the leo/... folder names now, I'm wondering if `extensions` was meant to be `external`?

**Comment by maphew** [comment]
Nice! Pruning external instead drops the wheel size almost 2mb, from 9.1 to to 7.3mb.

Is everything in external python modules? We can add them as optional in setup.py with `extras_require` [0],[1] (probably should be a separate issue if so)

[0]: http://setuptools.readthedocs.io/en/latest/setuptools.html#declaring-extras-optional-features-with-their-own-dependencies
[1]: https://stackoverflow.com/a/18879288/14420

**Comment by edreamleo** [comment]
On Sun, Oct 29, 2017 at 6:51 PM, matt wilkie <notifications@github.com>
wrote:

> Nice! Pruning external instead drops the wheel size almost 2mb, from 9.1
> to to 7.3mb.
>
> Is everything in external python modules?
>
​No.  It's a mixed bag.  I suggest including both the external and
extensions directories.

Edward

**Comment by maphew** [comment]
I will change to include both.

I spent some time looking/thinking about externals and extensions and how to use setup.py and pip to grab them instead of including. I concluded it's a dark passageway with twists and turns, needing time to map out and yielding uncertain benefit. So not for me right now.

For posterity, here's how they might be expressed in setup.py and then installed with `pip install leo[externals,extensions]`:

```
# Optional external modules (see `leo/externals`, `leo/extensions`)
extras_require = {
    'externals' : ['npyscreen','rope'],
    'extensions' : ['sh','patch==1.*']
    }
```

**Comment by maphew** [comment]
I pushed a change to setup.py that uses the source code management system for the version number in the pypi package instead of Leo's self reported number. The scm number is auto incremented from the latest release tag found in the repo. It looks like `leo-5.7.dev130+ge312165.d20171031`

For building, it requires the python module `setuptools_scm` installed and git exe be in PATH.

```
python setup.py bdist_wheel
...snip...
pip install dist\*
Processing .\dist\leo-5.7.dev130+ge312165.d20171031-py2.py3-none-any.whl
Installing collected packages: leo
Successfully installed leo-5.7.dev130+ge312165.d20171031
```

Leo still self reports using it's own system, so is out of step with the above. I haven't ventured into changing it as I don't understand the implications.  This could be considered scope creep and perhaps should be split into it's own issue. Especially if changing Leo's self-reporting is in anyway significant, or even just not desirable.

I ventured down this path because it didn't seem to me that "merely" adding a pip install-able setup is enough to justify a new release tag. (People who've installed 5.6 via other means won't care, but a release of 5.6.1 might make them think they should.)

```
b:\apps\leo-editor>leoc --version

Note: sys.stdout.encoding is not UTF-8
See: https://stackoverflow.com/questions/14109024

Leo 5.6, build 20171031115424, Tue Oct 31 11:54:24 CDT 2017
Not running from a git repo
Python 2.7.4, no gui!
Windows 7 AMD64 (build 6.1.7601) SP1
** isPython3: False
** caching enabled

Leo 5.6, build 20171031115424, Tue Oct 31 11:54:24 CDT 2017
```

**Comment by edreamleo** [comment]
On Wed, Nov 1, 2017 at 12:45 AM, matt wilkie <notifications@github.com>
wrote:

> I pushed a change to setup.py that uses the source code management system
> for the version number in the pypi package instead of Leo's self reported
> number.
>

​Thanks for this.  Do I need to take any action about this?

Edward

**Comment by maphew** [comment]
Ok, we're back to where things were before my big mess =)   Now `python setup.py bdist_wheel` will build a wheel package for uploading to pypi, and write a file with the current extracted-from-git version to an external file, `leo/version.py`:
~~~
# coding: utf-8
# file generated by setuptools_scm
# don't change, don't track in version control
version = '5.7.dev131+g9cdc135'
~~~


Is it feasible to call this file from leoVersion.py so that console reports the same version as the wheel package?

**Comment by tbnorth** [comment]
On Fri, 03 Nov 2017 05:27:14 +0000 (UTC)
matt wilkie <notifications@github.com> wrote:

> Ok, we're back to where things were before my big mess =)   Now
> `python setup.py bdist_wheel` will build a wheel package for
> uploading to pypi, and write a file with the current
> extracted-from-git version to an external file, `leo/version.py`: ~~~
> # coding: utf-8 # file generated by setuptools_scm # don't change,
> don't track in version control 
> version = '5.7.dev131+g9cdc135'

Is 131+g9cdc135 just a made up example?  Seems like it should be a
valid SHA1, like 131af9cdc135 or something.  Or is the 131+g meaning
131 commits since 5.7, now at *g*it commit 9cdc135?

> Is it feasible to call this file from leoVersion.py so that console
> reports the same version as the wheel package?

So currently we have

Leo 5.6, build 20171031115424, Tue Oct 31 11:54:24 CDT 2017
Git repo info: branch = master, commit = e312165e1934

and it would become

Leo 5.6.dev131+g9cdc135, build 20171031115424, Tue Oct 31 11:54:24 CDT 2017
Git repo info: branch = master, commit = e312165e1934

but how often will the PyPi package be build?  I don't think it should
be daily or anything, it should be made from a known good commit
periodically.  Unless PyPi supports nightly and version releases or
something.

So it seems the PyPi releases would always be X.Y.dev+nothing, so to
speak?

No doubt I'm missing details here.

Cheers -Terry

**Comment by maphew** [comment]
>
> > version = '5.7.dev131+g9cdc135'
>
> Is 131+g9cdc135 just a made up example? Seems like it should be a
> valid SHA1, like 131af9cdc135 or something. Or is the 131+g meaning
> 131 commits since 5.7, now at *g*it commit 9cdc135?
>

You have it pretty much. The default scheme
<https://github.com/pypa/setuptools_scm#default-versioning-scheme> is:
{next_version}.dev{distance}+{scm letter}{revision hash}
with 'next_version' being "last numbered tag release +1" and  the "g"
inherited from `git describe`. We could create a different template. I
don't think we need to be that verbose, and {next_version}.dev{distance}
would be fine. Anyone who wants a particular commit would be going straight
to the repo anyway right? (Unless the hash is useful/better for bug
hunting?).

I spent a little time trying to figure out the largely undocumented
template but didn't get very far. That doesn't say much; my programming
skills are meagre!


> but how often will the PyPi package be build? I don't think it should
> be daily or anything, it should be made from a known good commit
> periodically. Unless PyPi supports nightly and version releases or
> something.
>

I was envisioning milestones, minor releases, and/or just when someone
wants a new package published for whatever reason. (The present 'whatever
reason' being "install by pip works!" ;-)

Pip can install straight from Github so I wouldn't bother with more
frequent releases:

    pip install https://github.com/leo-editor/leo-editor/archive/master.zip

Matt

**Comment by tbnorth** [comment]
On Fri, 03 Nov 2017 20:02:47 +0000 (UTC)
matt wilkie <notifications@github.com> wrote:

> >
> > > version = '5.7.dev131+g9cdc135'
> >
> > Is 131+g9cdc135 just a made up example? Seems like it should be a
> > valid SHA1, like 131af9cdc135 or something. Or is the 131+g meaning
> > 131 commits since 5.7, now at *g*it commit 9cdc135?
> >
> 
> You have it pretty much. The default scheme
> <https://github.com/pypa/setuptools_scm#default-versioning-scheme> is:
> {next_version}.dev{distance}+{scm letter}{revision hash}
> with 'next_version' being "last numbered tag release +1" and  the "g"
> inherited from `git describe`. We could create a different template. I
> don't think we need to be that verbose, and
> {next_version}.dev{distance} would be fine. Anyone who wants a
> particular commit would be going straight to the repo anyway right?
> (Unless the hash is useful/better for bug hunting?).

I think we want to keep the hash in the log, and we also have the Leo
build number (timestamp), which again seems worth keeping.  So I'm kind
of wondering if there's room for all the variants.  Sp starting from
where we are:

Leo 5.6, build 20171031115424, Tue Oct 31 11:54:24 CDT 2017
Git repo info: branch = master, commit = 9cdc135643b6

what would you want it to look like?  Something like:

Leo 5.6.dev131+g9cdc1356, branch = master
build 20171031115424, Tue Oct 31 11:54:24 CDT 2017

perhaps?

I guess I don't mind, the information content's the same, plus distance.

See if Edward chimes in.

Cheers -Terry


> I spent a little time trying to figure out the largely undocumented
> template but didn't get very far. That doesn't say much; my
> programming skills are meagre!
> 
> 
> > but how often will the PyPi package be build? I don't think it
> > should be daily or anything, it should be made from a known good
> > commit periodically. Unless PyPi supports nightly and version
> > releases or something.
> >
> 
> I was envisioning milestones, minor releases, and/or just when someone
> wants a new package published for whatever reason. (The present
> 'whatever reason' being "install by pip works!" ;-)
> 
> Pip can install straight from Github so I wouldn't bother with more
> frequent releases:
> 
>     pip install
> https://github.com/leo-editor/leo-editor/archive/master.zip
> 
> Matt
> 
>

**Comment by maphew** [comment]
> what would you want it to look like? Something like:
>
> Leo 5.6.dev131+g9cdc1356, branch = master
> build 20171031115424, Tue Oct 31 11:54:24 CDT 2017
>

Build # looks redundant, it's just a compressed time stamp right?

That said my only real wish is to use the same number/phrase in Leo console
and the PyPi package. Details like whether commit hash and build number are
included are just {shrug}. They should be what's important to you and
Edward and the others who build the real features.

Matt

**Comment by tbnorth** [comment]
On Sat, 04 Nov 2017 05:06:37 +0000 (UTC)
matt wilkie <notifications@github.com> wrote:

> > what would you want it to look like? Something like:
> >
> > Leo 5.6.dev131+g9cdc1356, branch = master
> > build 20171031115424, Tue Oct 31 11:54:24 CDT 2017
> 
> Build # looks redundant, it's just a compressed time stamp right?

Yes, although if we wanted to trim one, I'd drop the expanded timestamp
in favor of the compressed, just because it's easier to copy / paste /
not misquote.  And searchable in git logs, seeing it's injected into
commit messages, if the developer's using the git hooks.

> That said my only real wish is to use the same number/phrase in Leo
> console and the PyPi package. Details like whether commit hash and
> build number are included are just {shrug}. They should be what's
> important to you and Edward and the others who build the real
> features.

Sure - the issue that might not be apparent is that while Edward tends
to tell people who want Leo to use git :-) I tend to tell them to use
zip, specifically
https://github.com/leo-editor/leo-editor/archive/master.zip

But here's the thing - a git repo. cannot include its own hash.
Because you have to create the commit to create the hash, and then you
can't add it to the commit without changing it, creating essentially a
different commit with a different hash.

So the 20171031115424 build number is the identity of a git commit for
people who aren't using git, e.g. people downloading the above zip.

Although the PyPi install *can* bundle a hash with the corresponding
commit's content, I'd still argue for "download a zip file" being
simpler for some users than "use pip", which I think is somewhat
simpler than "use git".

I'd be fine with

Leo 5.6.dev131+g9cdc1356
build: 20171031115424 branch: master

remembering the .zip version will be

Leo 5.6
build: 20171031115424

Really if the goal is consistency, perhaps the simplest is to have PyPi
packages display the build number, as above.  It has a 1:1 mapping to
commit hashes (except when two commits are made at the same second ;-)

It's in leo/core/commit_timestamp.json

Cheers -Terry

**Comment by maphew** [comment]
Thanks for the details Terry.

It's in leo/core/commit_timestamp.json
>

Which for me right now contains `"timestamp": "20171031115424"`, but latest
commit is 2017-11-02. Does this mean the json file is only updated when a
git tag is added?

Matt

**Comment by tbnorth** [comment]
On Sun, 05 Nov 2017 00:00:33 +0000 (UTC)
matt wilkie <notifications@github.com> wrote:

> Thanks for the details Terry.
> 
> It's in leo/core/commit_timestamp.json
> 
> Which for me right now contains `"timestamp": "20171031115424"`, but
> latest commit is 2017-11-02. Does this mean the json file is only
> updated when a git tag is added?
> 
> Matt

Only updated when committers install the git hooks... which I'm failing
to find in the docs.

Basically you copy 
  leo/extensions/hooks/commit-msg
and
  leo/extensions/hooks/pre-commit
to
  .git/hooks
i.e. the .git folder at the root of your Leo repo.

So it's fragile, in that you need to remember to do that after any
fresh clone and occasional committers might not do it, but nothing
breaks catastrophically when it's not done.

Cheers -Terry

**Comment by maphew** [comment]
Thanks Terry, I've now got the hooks installed* and am happy to see Leo's self-reported version and the packaging version closer to matching. 

So lets do away with importing yet another non-core module and just make PyPi version `{major}.{minor}.{build}`:

    5.6.20171031115424

IMO we don't need branch because it would be simpler overhead if we keep PyPi to mainline releases and continue to send developer/tester/itchy people go to github for the other stuff.


_\* Some clean-up in the extensions/hooks folder might needed. There are 2 pre-commit scripts in there, 1 bash and 1 python. The install_hooks.py installs the bash one. Also on commit there is a warning about LF replaced by CRLF in commit_timestamp.json. Dunno if that's important._

**Comment by tbnorth** [comment]
On Sun, 05 Nov 2017 21:13:58 +0000 (UTC)
matt wilkie <notifications@github.com> wrote:

> Thanks Terry, I've now got the hooks installed* and am happy to see
> Leo's self-reported version and the packaging version closer to
> matching. 
> 
> So lets do away with importing yet another non-core module and just
> make PyPi version `{major}.{minor}.{build}`:
> 
>     5.6.20171031115424
> 
> IMO we don't need branch because it would be simpler overhead if we
> keep PyPi to mainline releases and continue to send
> developer/tester/itchy people go to github for the other stuff.

There is a very small amount of information contained in the branch
name.  A commit / build number can occur on more than one
branch, so knowing that a user on on a branch, particularly a branch
other than master, conveys some information - telling them to pull and
see if it works now won't work if they're not pulling the branch you
expect them to pull.  So, very much edge case, but not completely
redundant.

> _\* Some clean-up in the extensions/hooks folder might needed. There
> are 2 pre-commit scripts in there, 1 bash and 1 python. The
> install_hooks.py installs the bash one. 

Heh, well I'm fine with deleting the Python versions Edward wrote and
keeping the Bash versions I wrote ;-)

Bash is git's default scripting language for hooks and pretty much
guaranteed to be installed and on the path when git's running - even in
Windows.  Can't say the same for Python.  Will let Edward chime in.

> Also on commit there is a
> warning about LF replaced by CRLF in commit_timestamp.json. Dunno if
> that's important._

I haven't seen / noticed that, might be a function of individual clones
and how the user sets the crlf translation options selected.

Cheers -Terry

**Comment by maphew** [comment]
Test package built with latest and uploaded
https://testpypi.python.org/pypi/leo/5.6.20171105131700

There's something weird with `long_description`. When built with Python 3.6 it only contains the first line from README.md. It's fine when built with py2.7. I don't know what to do about it.

The cropped description seems to be only difference. It installs and runs fine:

    pip install --index-url https://test.pypi.org/simple/ leo

**Comment by maphew** [comment]
The truncated long_description thing appears to be a bug in setuptools:
https://github.com/pypa/setuptools/issues/1126

**Comment by maphew** [comment]
I've dropped `setuptools_scm` and re-implemented just the wanted parts in Leo, by adding a new function `g.gitDescribe()` following the pattern of the existing git functions.

**Comment by maphew** [comment]
All the noise and conversation/debate over commit hash in the version/package name is moot. In spite of being PEP440 compliant it's not allowed on pypi.org at this time: https://github.com/pypa/pypi-legacy/issues/731

So I'm just using the `{tag}.dev{distance}` approach: --> `leo-5.7-dev220`.  
I can be argued into `{tag}.{datetime}` --> `leo-5.7.201711191059` but it's just so long/ugly to my eye. (follow up on mailing list)

**Comment by ranvik14** [comment]
> Pip can install straight from Github so I wouldn't bother with more frequent releases: 
> pip install https://github.com/leo-editor/leo-editor/archive/master.zip

I tried to do this in a freshly created venv - and - received the following error msg:

### 

(leo-57-dev) PS C:\Users\Viktor> pip list
DEPRECATION: The default format will switch to columns in the future. You can use --format=(legacy|columns) (or define a format=(legacy|columns) in your pip.conf under the [list] section) to disable this warning.
pip (9.0.1)
setuptools (28.8.0)
(leo-57-dev) PS C:\Users\Viktor> pip install https://github.com/leo-editor/leo-editor/archive/master.zip
Collecting https://github.com/leo-editor/leo-editor/archive/master.zip
  Downloading https://github.com/leo-editor/leo-editor/archive/master.zip
     / 11.3MB 2.2MB/s
    Complete output from command python setup.py egg_info:
    warning: pypandoc module not found, could not convert Markdown to RST
    Traceback (most recent call last):
      File "<string>", line 1, in <module>
      File "C:\Users\Viktor\AppData\Local\Temp\pip-2ct6h9fb-build\setup.py", line 69, in <module>
        long_description = read_md('README.md'),
    NameError: name 'read_md' is not defined

    ----------------------------------------
Command "python setup.py egg_info" failed with error code 1 in C:\Users\Viktor\AppData\Local\Temp\pip-2ct6h9fb-build\
(leo-57-dev) PS C:\Users\Viktor> python --version
Python 3.6.3
(leo-57-dev) PS C:\Users\Viktor>

###

Any explanation?

**Comment by ranvik14** [comment]
> PyPi compliant wheel packages are built and install successfully on Windows (7,10) and 
> Linux (Mint) with python 2.7 and 3.6.
>
> pip install --index-url https://test.pypi.org/simple/ leo

If I try to perform this installation approach I receive the following error msg, when starting Leo:

###
(leo-57-dev) PS C:\Users\Viktor> python --version
Python 3.6.3
(leo-57-dev) PS C:\Users\Viktor> pip list
DEPRECATION: The default format will switch to columns in the future. You can use --format=(legacy|columns) (or define a format=(legacy|columns) in your pip.conf under the [list] section) to disable this warning.
pip (9.0.1)
setuptools (28.8.0)
(leo-57-dev) PS C:\Users\Viktor> pip install --index-url https://test.pypi.org/simple/ leo
Collecting leo
  Using cached https://test-files.pythonhosted.org/packages/79/bb/444c27315a08d50ccbaa816f8b4b77275089178c7a45b9665fa7e0eaa3a0/leo-5.6.20171105131700-py2.py3-none-any.whl
Installing collected packages: leo
Successfully installed leo-5.6.20171105131700
(leo-57-dev) PS C:\Users\Viktor> pip list
DEPRECATION: The default format will switch to columns in the future. You can use --format=(legacy|columns) (or define a format=(legacy|columns) in your pip.conf under the [list] section) to disable this warning.
leo (5.6.20171105131700)
pip (9.0.1)
setuptools (28.8.0)
(leo-57-dev) PS C:\Users\Viktor> leoc
Traceback (most recent call last):
  File "C:\Users\Viktor\AppData\Local\Programs\Python\Python36\lib\runpy.py", line 193, in _run_module_as_main
    "__main__", mod_spec)
  File "C:\Users\Viktor\AppData\Local\Programs\Python\Python36\lib\runpy.py", line 85, in _run_code
    exec(code, run_globals)
  File "C:\Users\Viktor\leo-57-dev\Scripts\leoc.exe\__main__.py", line 5, in <module>
  File "c:\users\viktor\leo-57-dev\lib\site-packages\leo\core\runLeo.py", line 26, in <module>
    import leo.core.leoApp as leoApp
  File "c:\users\viktor\leo-57-dev\lib\site-packages\leo\core\leoApp.py", line 31, in <module>
    import leo.core.editpane.editpane  # for the command creation side effect
  File "c:\users\viktor\leo-57-dev\lib\site-packages\leo\core\editpane\editpane.py", line 26, in <module>
    from leo.core.editpane.clicky_splitter import ClickySplitter
  File "c:\users\viktor\leo-57-dev\lib\site-packages\leo\core\editpane\clicky_splitter.py", line 13, in <module>
    class ClickySplitterHandle(QtWidgets.QSplitterHandle):
AttributeError: 'NoneType' object has no attribute 'QSplitterHandle'
(leo-57-dev) PS C:\Users\Viktor>

###

Any explanation?

**Comment by maphew** [comment]
>(leo-57-dev) PS C:\Users\Viktor> pip install --index-url
https://test.pypi.org/simple/ leo
>Collecting leo
>Using cached https://test-files.pythonhosted.org/packages/79/bb/
444c27315a08d50ccbaa816f8b4b77275089178c7a45b9665fa7e0eaa3a0
/leo-5.6.20171105131700-py2.py3-none-any.whl

Try `pip install --pre ..` and/or `pip install --no-cache-dir`. It's using
a cached 5.6 package and not the latest, but one I expected to work anyway.
Do you have pyqt5 or pyqt4 installed?

Please create a new Issue, this one is getting too multi-headed. Thanks for
testing!

matt

**Comment by tbnorth** [comment]
On Sun, 19 Nov 2017 13:32:08 +0000 (UTC)
ranvik14 <notifications@github.com> wrote:

>     class ClickySplitterHandle(QtWidgets.QSplitterHandle):
> AttributeError: 'NoneType' object has no attribute 'QSplitterHandle'

I think it happens when Qt fails to load; which could verified by just
entering

python

and then

from PyQt5 import QtCore
from PyQt4 import QtCore

if both of those fail, Qt can't be located.

I tried fixing this by moving the code in question into a scope only
entered by Leo when it's importing Qt, but keep forgetting that failing
to import Qt just returns None for the modules, instead of raising an
ImportError.  So I'll modify the code to check for that condition.

Cheers -Terry

**Comment by tbnorth** [comment]
On Sun, 19 Nov 2017 08:27:11 -0600
Terry Brown <terrynbrown@gmail.com> wrote:

> On Sun, 19 Nov 2017 13:32:08 +0000 (UTC)
> ranvik14 <notifications@github.com> wrote:
> 
> >     class ClickySplitterHandle(QtWidgets.QSplitterHandle):
> > AttributeError: 'NoneType' object has no attribute
> > 'QSplitterHandle'  

I think 32d6815 will prevent the reported error, but I still think it's
just a symptom of Qt not loading, so you'll get a hopefully more
diagnostic error now.

Cheers -Terry

> I think it happens when Qt fails to load; which could verified by just
> entering
>
> python
> 
> and then
> 
> from PyQt5 import QtCore
> from PyQt4 import QtCore
> 
> if both of those fail, Qt can't be located.
> 
> I tried fixing this by moving the code in question into a scope only
> entered by Leo when it's importing Qt, but keep forgetting that
> failing to import Qt just returns None for the modules, instead of
> raising an ImportError.  So I'll modify the code to check for that
> condition.
> 
> Cheers -Terry
>

**Comment by ranvik14** [comment]
Hello Terry,

2017-11-19 15:27 GMT+01:00 Terry Brown <notifications@github.com>:

> On Sun, 19 Nov 2017 13:32:08 +0000 (UTC)
> ranvik14 <notifications@github.com> wrote:
>
> > class ClickySplitterHandle(QtWidgets.QSplitterHandle):
> > AttributeError: 'NoneType' object has no attribute 'QSplitterHandle'
>
> I think it happens when Qt fails to load; which could verified by just
> entering
>
> python
>
> and then
>
> from PyQt5 import QtCore
> from PyQt4 import QtCore
>
> if both of those fail, Qt can't be located.
>

I'll verify that as soon as possible & report back ...

I tried fixing this by moving the code in question into a scope only
> entered by Leo when it's importing Qt, but keep forgetting that failing
> to import Qt just returns None for the modules, instead of raising an
> ImportError. So I'll modify the code to check for that condition.
>

With kind regards,

Viktor

**Comment by ranvik14** [comment]
> Pip can install straight from Github so I wouldn't bother with more frequent releases:
> pip install https://github.com/leo-editor/leo-editor/archive/master.zip

I tried to do this again in a freshly created venv, now I installed PyQt5 before as well - but - I *still* received the following error msg:

###

Windows PowerShell
Copyright (C) Microsoft Corporation. Alle Rechte vorbehalten.

PS C:\Users\Viktor> cd leo-57-dev
PS C:\Users\Viktor\leo-57-dev> ./Scripts/Activate.ps1
(leo-57-dev) PS C:\Users\Viktor\leo-57-dev> pip list
DEPRECATION: The default format will switch to columns in the future. You can use --format=(legacy|columns) (or define a format=(legacy|columns) in your pip.conf under the [list] section) to disable this warning.
pip (9.0.1)
setuptools (28.8.0)
(leo-57-dev) PS C:\Users\Viktor\leo-57-dev> pip install pyqt5
Collecting pyqt5
  Downloading PyQt5-5.9.1-5.9.2-cp35.cp36.cp37-none-win_amd64.whl (77.4MB)
    100% |████████████████████████████████| 77.4MB 13kB/s
Collecting sip<4.20,>=4.19.4 (from pyqt5)
  Downloading sip-4.19.5-cp36-none-win_amd64.whl (51kB)
    100% |████████████████████████████████| 51kB 1.1MB/s
Installing collected packages: sip, pyqt5
Successfully installed pyqt5-5.9.1 sip-4.19.5
(leo-57-dev) PS C:\Users\Viktor\leo-57-dev> pip list
DEPRECATION: The default format will switch to columns in the future. You can use --format=(legacy|columns) (or define a format=(legacy|columns) in your pip.conf under the [list] section) to disable this warning.
pip (9.0.1)
PyQt5 (5.9.1)
setuptools (28.8.0)
sip (4.19.5)
(leo-57-dev) PS C:\Users\Viktor\leo-57-dev> pip install https://github.com/leo-editor/leo-editor/archive/master.zip
Collecting https://github.com/leo-editor/leo-editor/archive/master.zip
  Downloading https://github.com/leo-editor/leo-editor/archive/master.zip
     / 11.3MB 2.2MB/s
    Complete output from command python setup.py egg_info:
    warning: pypandoc module not found, could not convert Markdown to RST
    execGitCommand not found: C:\Users\Viktor\AppData\Local\Temp\pip-vsmniz0n-build\.git
    Traceback (most recent call last):
      File "<string>", line 1, in <module>
      File "C:\Users\Viktor\AppData\Local\Temp\pip-vsmniz0n-build\setup.py", line 80, in <module>
        version = git_version(),
      File "C:\Users\Viktor\AppData\Local\Temp\pip-vsmniz0n-build\setup.py", line 33, in git_version
        tag, distance, commit = g.gitDescribe(root)
      File "C:\Users\Viktor\AppData\Local\Temp\pip-vsmniz0n-build\leo\core\leoGlobals.py", line 4272, in gitDescribe
        tag, distance, commit = describe[0].split('-')
    TypeError: 'NoneType' object is not subscriptable

    ----------------------------------------
Command "python setup.py egg_info" failed with error code 1 in C:\Users\Viktor\AppData\Local\Temp\pip-vsmniz0n-build\
(leo-57-dev) PS C:\Users\Viktor\leo-57-dev>

###

Any explanation?

With kind regards,

Viktor

**Comment by ranvik14** [comment]
> PyPi compliant wheel packages are built and install successfully on Windows (7,10) and
> Linux (Mint) with python 2.7 and 3.6.
>
> pip install --index-url https://test.pypi.org/simple/ leo

After I did install PyQt5 before, this method did work for me. - Here are the logs:

###

(leo-57-dev) PS C:\Users\Viktor\leo-57-dev> python --version
Python 3.6.3
(leo-57-dev) PS C:\Users\Viktor\leo-57-dev> pip list
DEPRECATION: The default format will switch to columns in the future. You can use --format=(legacy|columns) (or define a format=(legacy|columns) in your pip.conf under the [list] section) to disable this warning.
pip (9.0.1)
PyQt5 (5.9.1)
setuptools (28.8.0)
sip (4.19.5)
(leo-57-dev) PS C:\Users\Viktor\leo-57-dev> pip install --index-url https://test.pypi.org/simple/ leo
Collecting leo
  Using cached https://test-files.pythonhosted.org/packages/79/bb/444c27315a08d50ccbaa816f8b4b77275089178c7a45b9665fa7e0eaa3a0/leo-5.6.20171105131700-py2.py3-none-any.whl
Installing collected packages: leo
Successfully installed leo-5.6.20171105131700
(leo-57-dev) PS C:\Users\Viktor\leo-57-dev> pip list
DEPRECATION: The default format will switch to columns in the future. You can use --format=(legacy|columns) (or define a format=(legacy|columns) in your pip.conf under the [list] section) to disable this warning.
leo (5.6.20171105131700)
pip (9.0.1)
PyQt5 (5.9.1)
setuptools (28.8.0)
sip (4.19.5)
(leo-57-dev) PS C:\Users\Viktor\leo-57-dev> leoc

reading settings in C:\users\viktor\leo-57-dev\lib\site-packages\leo\config\leoSettings.leo
reading settings in C:\Users\Viktor\.leo\myLeoSettings.leo

Leo 5.6, build 20171105131700, Sun Nov  5 13:17:00 PST 2017
Not running from a git repo
Python 3.6.3, PyQt version 5.9.2
Windows 10 AMD64 (build 10.0.16299) SP0
** isPython3: True
** caching enabled

Using default leo file name:
C:\Users\Viktor\.leo\workbook.leo
loadOnePlugin: can not load enabled plugin: leo.plugins.viewrendered
reading settings in C:\users\viktor\leo-57-dev\lib\site-packages\leo\doc\CheatSheet.leo
loadOnePlugin: can not load enabled plugin: leo.plugins.viewrendered
wrote recent file: C:\Users\Viktor\.leo\.leoRecentFiles.txt
(leo-57-dev) PS C:\Users\Viktor\leo-57-dev>

###

Thanks a lot! - Looking forward to getting Leo from PyPI ...

With kind regards,

Viktor

**Comment by tbnorth** [comment]
On Mon, 20 Nov 2017 06:10:44 +0000 (UTC)
ranvik14 <notifications@github.com> wrote:

> I tried to do this again in a freshly created venv, now I installed
> PyQt5 before as well - but - I *still* received the following error
> msg:

But a different error, I think?

> "C:\Users\Viktor\AppData\Local\Temp\pip-vsmniz0n-build\leo\core\leoGlobals.py",
> line 4272, in gitDescribe tag, distance, commit =
> describe[0].split('-') TypeError: 'NoneType' object is not
> subscriptable

One Matt probably wants to take a look at :-)

Cheers -Terry

**Comment by maphew** [comment]
opening post has been updated to reflect current status.

**Comment by ranvik14** [comment]
An installation into a new virtual environment succeeded. - Here's the log

<Log>

(leo-57-dev) PS C:\Users\Viktor> python -m venv leo-57-dev
(leo-57-dev) PS C:\Users\Viktor> pip list
DEPRECATION: The default format will switch to columns in the future. You can use --format=(legacy|columns) (or define a format=(legacy|columns) in your pip.conf under the [list] section) to disable this warning.
pip (9.0.1)
setuptools (28.8.0)
(leo-57-dev) PS C:\Users\Viktor> pip install PyQt5
Collecting PyQt5
  Using cached PyQt5-5.9.2-5.9.3-cp35.cp36.cp37-none-win_amd64.whl
Collecting sip<4.20,>=4.19.4 (from PyQt5)
  Using cached sip-4.19.6-cp36-none-win_amd64.whl
Installing collected packages: sip, PyQt5
Successfully installed PyQt5-5.9.2 sip-4.19.6
(leo-57-dev) PS C:\Users\Viktor> pip list
DEPRECATION: The default format will switch to columns in the future. You can use --format=(legacy|columns) (or define a format=(legacy|columns) in your pip.conf under the [list] section) to disable this warning.
pip (9.0.1)
PyQt5 (5.9.2)
(leo-57-dev) PS C:\Users\Viktor> pip install --no-cache-dir Leo
Collecting Leo
  Downloading leo-5.7.dev344-py2.py3-none-any.whl (9.2MB)
    100% |████████████████████████████████| 9.2MB 2.2MB/s
Collecting docutils (from Leo)
  Downloading docutils-0.14-py3-none-any.whl (543kB)
    100% |████████████████████████████████| 552kB 2.2MB/s
Collecting sphinx (from Leo)
  Downloading Sphinx-1.6.5-py2.py3-none-any.whl (1.9MB)
    100% |████████████████████████████████| 1.9MB 1.6MB/s
Requirement already satisfied: PyQt5; python_version >= "3.0" in c:\users\viktor\leo-57-dev\lib\site-packages (from Leo)
Collecting Pygments>=2.0 (from sphinx->Leo)
  Downloading Pygments-2.2.0-py2.py3-none-any.whl (841kB)
    100% |████████████████████████████████| 849kB 2.2MB/s
Collecting babel!=2.0,>=1.3 (from sphinx->Leo)
  Downloading Babel-2.5.1-py2.py3-none-any.whl (6.8MB)
    100% |████████████████████████████████| 6.8MB 1.6MB/s
Collecting colorama>=0.3.5; sys_platform == "win32" (from sphinx->Leo)
  Downloading colorama-0.3.9-py2.py3-none-any.whl
Collecting snowballstemmer>=1.1 (from sphinx->Leo)
  Downloading snowballstemmer-1.2.1-py2.py3-none-any.whl (64kB)
    100% |████████████████████████████████| 71kB 2.3MB/s
Collecting alabaster<0.8,>=0.7 (from sphinx->Leo)
  Downloading alabaster-0.7.10-py2.py3-none-any.whl
Collecting imagesize (from sphinx->Leo)
  Downloading imagesize-0.7.1-py2.py3-none-any.whl
Collecting requests>=2.0.0 (from sphinx->Leo)
  Downloading requests-2.18.4-py2.py3-none-any.whl (88kB)
    100% |████████████████████████████████| 92kB 2.9MB/s
Collecting sphinxcontrib-websupport (from sphinx->Leo)
  Downloading sphinxcontrib_websupport-1.0.1-py2.py3-none-any.whl
Collecting six>=1.5 (from sphinx->Leo)
  Downloading six-1.11.0-py2.py3-none-any.whl
Requirement already satisfied: setuptools in c:\users\viktor\leo-57-dev\lib\site-packages (from sphinx->Leo)
Collecting Jinja2>=2.3 (from sphinx->Leo)
  Downloading Jinja2-2.10-py2.py3-none-any.whl (126kB)
    100% |████████████████████████████████| 133kB 2.2MB/s
Requirement already satisfied: sip<4.20,>=4.19.4 in c:\users\viktor\leo-57-dev\lib\site-packages (from PyQt5; python_version >= "3.0"->Leo)
Collecting pytz>=0a (from babel!=2.0,>=1.3->sphinx->Leo)
  Downloading pytz-2017.3-py2.py3-none-any.whl (511kB)
    100% |████████████████████████████████| 512kB 2.2MB/s
Collecting idna<2.7,>=2.5 (from requests>=2.0.0->sphinx->Leo)
  Downloading idna-2.6-py2.py3-none-any.whl (56kB)
    100% |████████████████████████████████| 61kB 3.9MB/s
Collecting urllib3<1.23,>=1.21.1 (from requests>=2.0.0->sphinx->Leo)
  Downloading urllib3-1.22-py2.py3-none-any.whl (132kB)
    100% |████████████████████████████████| 133kB 2.2MB/s
Collecting certifi>=2017.4.17 (from requests>=2.0.0->sphinx->Leo)
  Downloading certifi-2017.11.5-py2.py3-none-any.whl (330kB)
    100% |████████████████████████████████| 337kB 1.6MB/s
Collecting chardet<3.1.0,>=3.0.2 (from requests>=2.0.0->sphinx->Leo)
  Downloading chardet-3.0.4-py2.py3-none-any.whl (133kB)
    100% |████████████████████████████████| 143kB 1.6MB/s
Collecting MarkupSafe>=0.23 (from Jinja2>=2.3->sphinx->Leo)
  Downloading MarkupSafe-1.0.tar.gz
Installing collected packages: docutils, Pygments, pytz, babel, colorama, snowballstemmer, alabaster, imagesize, idna, urllib3, certifi, chardet, requests, sphinxcontrib-websupport, six, MarkupSafe, Jinja2, sphinx, Leo
  Running setup.py install for MarkupSafe ... done
Successfully installed Jinja2-2.10 Leo-5.7.dev344 MarkupSafe-1.0 Pygments-2.2.0 alabaster-0.7.10 babel-2.5.1 certifi-2017.11.5 chardet-3.0.4 colorama-0.3.9 docutils-0.14 idna-2.6 imagesize-0.7.1 pytz-2017.3 requests-2.18.4 six-1.11.0 snowballstemmer-1.2.1 sphinx-1.6.5 sphinxcontrib-websupport-1.0.1 urllib3-1.22
(leo-57-dev) PS C:\Users\Viktor> pip list
DEPRECATION: The default format will switch to columns in the future. You can use --format=(legacy|columns) (or define a format=(legacy|columns) in your pip.conf under the [list] section) to disable this warning.
alabaster (0.7.10)
Babel (2.5.1)
certifi (2017.11.5)
chardet (3.0.4)
colorama (0.3.9)
docutils (0.14)
idna (2.6)
imagesize (0.7.1)
Jinja2 (2.10)
leo (5.7.dev344)
MarkupSafe (1.0)
pip (9.0.1)
Pygments (2.2.0)
PyQt5 (5.9.2)
pytz (2017.3)
requests (2.18.4)
setuptools (28.8.0)
sip (4.19.6)
six (1.11.0)
snowballstemmer (1.2.1)
Sphinx (1.6.5)
sphinxcontrib-websupport (1.0.1)
urllib3 (1.22)
(leo-57-dev) PS C:\Users\Viktor>

</Log>

For a Python 3 in Windows 10 this issue has been fully resolved. - Thanks!

**Comment by ranvik14** [comment]
@Matt: I provided my feedback there. - In short: For a Python 3 in Windows
10 environment everything looks good. - Thanks!


2017-11-30 1:13 GMT+01:00 matt wilkie <notifications@github.com>:

**Comment by maphew** [comment]
@edreamleo: Edward I think this is ready for testing by you. I updated LeoDist and the opening comment to this issue accordingly with the recipe.

(Anyone can test the build recipe, it's only `twine upload ...` that requires something extra. Speaking of which any of "the usual suspects" can be added to the package maintainer's role, pending Edward's agreement of course.)

**Comment by edreamleo** [comment]
On Sun, Dec 10, 2017 at 12:30 AM, matt wilkie <notifications@github.com>
wrote:

> @edreamleo <https://github.com/edreamleo>: Edward I think this is ready
> for testing by you. I updated LeoDist and the opening comment to this issue
> accordingly with the recipe.
>
​I'll do this when I can. It may be a few weeks.

Edward

**Comment by maphew** [comment]
Updated top post: bigger headings to more clearly separate the various sections. Fixed name of `leo-console` command.

**Comment by edreamleo** [comment]
I am going to close this:

- Matt's notes in leoDist.py are excellent: clear and complete.
- Uploading a Leo wheel must wait until after Leo 5.7 final goes out the door.
- Matt can edit this long discussion, or not, as he pleases.
